src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,236 @@
+package se.liu.ida.hefquin.engine.queryplan.executable.impl.ops;
+
+import java.util.List;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.HashSet;
+
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import org.apache.jena.sparql.engine.binding.BindingBuilder;
+
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.Triple;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingImpl;
+import se.liu.ida.hefquin.engine.data.impl.TripleImpl;
+import se.liu.ida.hefquin.engine.federation.TPFServer;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessException;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessManager;
+import se.liu.ida.hefquin.engine.federation.access.TPFRequest;
+import se.liu.ida.hefquin.engine.federation.access.TPFResponse;
+import se.liu.ida.hefquin.engine.federation.access.TriplePatternRequest;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.TPFRequestImpl;
+import se.liu.ida.hefquin.engine.federation.access.impl.response.TPFResponseImpl;
+import se.liu.ida.hefquin.engine.federation.access.utils.FederationAccessUtils;
+import se.liu.ida.hefquin.engine.query.BGP;
+import se.liu.ida.hefquin.engine.query.SPARQLGraphPattern;
+import se.liu.ida.hefquin.engine.query.SPARQLGroupPattern;
+import se.liu.ida.hefquin.engine.query.SPARQLUnionPattern;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+
+/**
+ * Implementation of an operator to request a (complete) TPF from a TPF server.
+ * This implementation handles pagination of the TPF; that is, it requests all
+ * the pages, one after another.
+ */
+public class ExecOpRequestTPFatTPFServerWithTranslation extends ExecOpGenericTriplePatternRequestWithTPF<TPFServer>
+{
+	public ExecOpRequestTPFatTPFServerWithTranslation( final TriplePatternRequest req, final TPFServer fm ) {
+		super( req, fm );
+	}
+
+	@Override
+	protected TPFResponse performPageRequest( final TPFRequest req,
+	                                          final FederationAccessManager fedAccessMgr )
+			throws FederationAccessException
+	{
+		final SPARQLGraphPattern reqTranslation = fm.getVocabularyMapping().translateTriplePattern(req.getQueryPattern());
+		final List<Triple> resList = new ArrayList<>();
+		if(reqTranslation instanceof TriplePattern) {
+			resList.addAll(handleTriplePattern((TriplePattern) reqTranslation, fedAccessMgr));
+		}
+		
+		else if(reqTranslation instanceof SPARQLUnionPattern) {
+			resList.addAll(handleUnionPattern(((SPARQLUnionPattern) reqTranslation), fedAccessMgr));
+		}
+		
+		else if(reqTranslation instanceof SPARQLGroupPattern) {
```

Comment:
I don't think that your translation method works for arbitrary SPARQL queries, right? So, I'm not sure that this case should be supported here. Notice that if the translation of a triple pattern produces a SPARQLGroupPattern, we have to support that, but we do not know how to translate a SPARQLGroupPattern (right?)

Reply:
SPARQLGroupPatterns are used to create intersections when all memebers of the intersection aren't triplepatterns (so BGP can't be used). So there could be a SPARQLGroupPattern in the translation